stdlib/internal/SwiftSyntax/SyntaxData.swift

```diff
@@ -0,0 +1,196 @@
+//===-------------------- SyntaxData.swift - Syntax Data ------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+/// A unique identifier for a node in the tree.
+typealias NodeIdentifier = [Int]
+
+/// SyntaxData is the underlying storage for each Syntax node.
+/// It's modelled as an array that stores and caches a SyntaxData for each raw
+/// syntax node in its layout. It is up to the specific Syntax nodes to maintain
+/// the correct layout of their SyntaxData backing stores.
+///
+/// SyntaxData is an implementation detail, and should not be exposed to clients
+/// of libSyntax.
+///
+/// The following relationships are preserved:
+///   parent.cachedChild(at: indexInParent) === self
+///   raw.layout.count == childCaches.count
+///   pathToRoot.first === indexInParent
+final class SyntaxData: Equatable {
+
+  let raw: RawSyntax
+  let indexInParent: Int
+  weak var parent: SyntaxData?
```

Comment:
Can we end up in a situation where the parent goes away without the syntax data being "invalid" in some send, or could this be `unowned`?